### PR TITLE
Add ability for users to allow tuple drops upon backpressure detection

### PR DESCRIPTION
**Pull request by srkukarni**

### CODE REVIEW

File: heron/common/src/cpp/network/client.cpp [hunk 1]
```diff
@@ -218,3 +218,8 @@ void Client::StartBackPressureConnectionCb(Connection* conn) {
 void Client::StopBackPressureConnectionCb(Connection* conn) {
   // Nothing to be done here. Should be handled by inheritors if they care about backpressure
 }
+
+bool Client::HasCausedBackPressure() const {
+  Connection* conn = static_cast<Connection*>(conn_);
```
Comment (nwangtw) [inline heron/common/src/cpp/network/client.cpp]: Do we need to check state&pointer here? Can conn_ be NULL in some cases/states?

File: heron/stmgr/src/cpp/manager/instance-server.cpp [hunk 2]
```diff
@@ -356,6 +358,10 @@ void InstanceServer::DrainTupleStream(proto::stmgr::TupleStreamMessage* _message
   TaskIdInstanceDataMap::iterator iter = instance_info_.find(task_id);
   if (iter == instance_info_.end() || iter->second->conn_ == NULL) {
     LOG(ERROR) << "task_id " << task_id << " has not yet connected to us. Dropping...";
+  } else if (droptuples_upon_backpressure_ && iter->second->conn_->hasCausedBackPressure()) {
+    LOG(ERROR) << "task_id " << task_id << " is causing backpressure, so Dropping tuples to him";
```
Comment (nwangtw) [inline heron/stmgr/src/cpp/manager/instance-server.cpp]: suggest to add "and droptuples_upon_backpressure is set to true" in log.

Also, how many logs are we talking about here?

File: heron/stmgr/src/cpp/manager/instance-server.cpp [hunk 3]
```diff
@@ -356,6 +358,10 @@ void InstanceServer::DrainTupleStream(proto::stmgr::TupleStreamMessage* _message
...
+    // Ideally we would have counted the numbers lost, but we cannot since the tuples are in
```
Comment (nwangtw) [inline heron/stmgr/src/cpp/manager/instance-server.cpp]: Do we have any information for the package we can log? Like time window, etc? Sorry if this is a silly question.

File: heron/stmgr/src/cpp/manager/instance-server.cpp [hunk 4]
```diff
@@ -371,16 +377,24 @@ void InstanceServer::SendToInstance2(sp_int32 _task_id,
 void InstanceServer::DrainTupleSet(sp_int32 _task_id,
                                 proto::system::HeronTupleSet2* _message) {
   TaskIdInstanceDataMap::iterator iter = instance_info_.find(_task_id);
-  if (iter == instance_info_.end() || iter->second->conn_ == NULL) {
-    LOG(ERROR) << "task_id " << _task_id << " has not yet connected to us. Dropping...";
-    if (_message->has_control()) {
+  if (iter == instance_info_.end() || iter->second->conn_ == NULL
+      || (droptuples_upon_backpressure_ && iter->second->conn_->hasCausedBackPressure())) {
+    LOG(ERROR) << "task_id " << _task_id
+               << " has not yet connected to us or is not keeping up, Dropping...";
```
Comment (nwangtw) [inline heron/stmgr/src/cpp/manager/instance-server.cpp]: same, suggest to add "and droptuples_upon_backpressure is set to true" in log.

File: heron/stmgr/src/cpp/manager/instance-server.cpp [hunk 5]
```diff
@@ -463,6 +483,13 @@ void InstanceServer::StopBackPressureConnectionCb(Connection* _connection) {
   sp_string instance_name = GetInstanceName(_connection);
   CHECK_NE(instance_name, "");
 
+  LOG(INFO) << "We don't observe back pressure now on sending data to instance " << instance_name;
```
Comment (nwangtw) [inline heron/stmgr/src/cpp/manager/instance-server.cpp]: I am just wondering if this should be a WARNING log. Not a blocker.

File: heron/stmgr/src/cpp/manager/instance-server.cpp [hunk 6]
```diff
@@ -463,6 +483,13 @@ void InstanceServer::StopBackPressureConnectionCb(Connection* _connection) {
...
+
+  if (droptuples_upon_backpressure_) {
```
Comment (nwangtw) [inline heron/stmgr/src/cpp/manager/instance-server.cpp]: Suggest to add a log here like "droptuples_upon_backpressure is set to true, hence no backpressure mechanism is initialized"

File: heron/stmgr/src/cpp/manager/stmgr-client.cpp [hunk 7]
```diff
@@ -192,6 +194,12 @@ bool StMgrClient::SendTupleStreamMessage(proto::stmgr::TupleStreamMessage& _msg)
   if (IsConnected()) {
     SendMessage(_msg);
     return true;
+  } else if (droptuples_upon_backpressure_ && HasCausedBackPressure()) {
+    if (++ndropped_messages_ % 100 == 0) {
+      LOG(INFO) << "Dropping " << ndropped_messages_ << "th tuple message to stmgr "
```
Comment (nwangtw) [inline heron/stmgr/src/cpp/manager/stmgr-client.cpp]: The drop count is for both not connected and backpressure, so the log could be misleading. Just a concern to me, not a real issue. Your call.

File: heron/stmgr/src/cpp/manager/stmgr-client.cpp [hunk 8]
```diff
@@ -192,6 +194,12 @@ bool StMgrClient::SendTupleStreamMessage(proto::stmgr::TupleStreamMessage& _msg)
   if (IsConnected()) {
     SendMessage(_msg);
```
Comment (nwangtw) [inline heron/stmgr/src/cpp/manager/stmgr-client.cpp]: Just to confirm. The drop upon backpressure logic is handled in the "(!IsConnected())" branch. Is it expected? Do we need it in the "(IsConnected)" branch?

File: heron/stmgr/src/cpp/manager/stmgr-client.cpp [hunk 9]
```diff
@@ -211,13 +219,17 @@ void StMgrClient::StartBackPressureConnectionCb(Connection* _connection) {
   // Ask the StMgrServer to stop consuming. The client does
   // not consume anything
 
-  client_manager_->StartBackPressureOnServer(other_stmgr_id_);
+  if (!droptuples_upon_backpressure_) {
+    client_manager_->StartBackPressureOnServer(other_stmgr_id_);
```
Comment (nwangtw) [inline heron/stmgr/src/cpp/manager/stmgr-client.cpp]: I don't know this part of code yet. Just to confirm, the StartBackPressureOnServer() call doesn't affect the HasCausedBackpressure() result, right?

**Comment by srkukarni** [comment]
@nwangtw This flag makes sense even in ATLEAST ONCE because the tuples will time out and will retry later

**Comment by nwangtw** [comment]
Got it. Thanks!

File: heron/common/src/cpp/network/client.cpp (hunk 1 above)
Comment (srkukarni) [reply heron/common/src/cpp/network/client.cpp]: Not really, but I've added some checks just for safety

File: heron/stmgr/src/cpp/manager/instance-server.cpp (hunk 2 above)
Comment (srkukarni) [reply heron/stmgr/src/cpp/manager/instance-server.cpp]: Expanded comment and also log every so often

File: heron/stmgr/src/cpp/manager/instance-server.cpp (hunk 3 above)
Comment (srkukarni) [reply heron/stmgr/src/cpp/manager/instance-server.cpp]: Addressed by adding bytes dropped

File: heron/stmgr/src/cpp/manager/instance-server.cpp (hunk 4 above)
Comment (srkukarni) [reply heron/stmgr/src/cpp/manager/instance-server.cpp]: Added

File: heron/stmgr/src/cpp/manager/instance-server.cpp (hunk 5 above)
Comment (srkukarni) [reply heron/stmgr/src/cpp/manager/instance-server.cpp]: Sounds good

File: heron/stmgr/src/cpp/manager/instance-server.cpp (hunk 6 above)
Comment (srkukarni) [reply heron/stmgr/src/cpp/manager/instance-server.cpp]: Added

File: heron/stmgr/src/cpp/manager/stmgr-client.cpp (hunk 8 above)
Comment (srkukarni) [reply heron/stmgr/src/cpp/manager/stmgr-client.cpp]: Redid the logic

File: heron/stmgr/src/cpp/manager/stmgr-client.cpp (hunk 7 above)
Comment (srkukarni) [reply heron/stmgr/src/cpp/manager/stmgr-client.cpp]: I think as long as we are counting drops, we should be ok

File: heron/stmgr/src/cpp/manager/stmgr-client.cpp (hunk 9 above)
Comment (srkukarni) [reply heron/stmgr/src/cpp/manager/stmgr-client.cpp]: Nopes

**Comment by nwangtw** [comment]
Cool. Overall LGTM.

One last comment is to add the config in storm compatibility as well to keep them in sync.

**Comment by srkukarni** [comment]
@nwangtw 1. Does Storm have an equivalent call?

**Comment by nwangtw** [comment]
Good question. 

I thought that the config as well and the translation should be added to the storm layer. Or maybe it shouldn't since the storm API shouldn't have the new features? I think I am ok either way.

**Comment by srkukarni** [comment]
Unless there is an equivalent storm api, there is no point adding this flag in the storm layer. As you mentioned storm layer is strictly for emulating the storm api, not adding func to it.

**Comment by nwangtw** [comment]
Make sense. Thanks.

**Pull request merged**